src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -209,39 +209,57 @@ private static ReadOnlyMemory<byte> GetCoseAlgorithmFromProtectedHeaders(CoseHea
+        // If we Slip: The caller did not specify a COSE Algorithm, we will write the header for them, rather than throw.
+        private static void ValidateOrSlipAlgorithmHeader(
+            CoseHeaderMap protectedHeaders,
+            CoseHeaderMap unprotectedHeaders,
+            KeyType keyType,
+            HashAlgorithmName hashAlgorithm,
+            out int? headerValueToSlip)
         {
-            int algHeaderValue;
+            int algHeaderValue = GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(keyType, hashAlgorithm);
 
             if (protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
             {
-                int? algHeaderValueNullable = DecodeCoseAlgorithmHeader(encodedAlg);
-                Debug.Assert(algHeaderValueNullable.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
-                algHeaderValue = algHeaderValueNullable.Value;
-
-                HashAlgorithmName expectedHashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algHeaderValue, keyType);
-
-                if (expectedHashAlgorithm != hashAlgorithm)
-                {
-                    throw new CryptographicException(SR.Format(SR.Sign1SignHashAlgorithmDoesNotMatchSpecifiedAlg, hashAlgorithm.Name, algHeaderValue));
-                }
+                ValidateAlgorithmHeader(encodedAlg, algHeaderValue, keyType, hashAlgorithm);
+                headerValueToSlip = null;
+                return;
             }
 
             if (unprotectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out _))
             {
-                throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+                throw new CryptographicException(SR.Sign1SignAlgMustBeProtected);
             }
 
-            algHeaderValue = GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(keyType, hashAlgorithm);
-            protectedHeaders.SetValue(CoseHeaderLabel.Algorithm, algHeaderValue);
+            headerValueToSlip = algHeaderValue;
+
+            static void ValidateAlgorithmHeader(ReadOnlyMemory<byte> encodedAlg, int expectedAlg, KeyType keyType, HashAlgorithmName hashAlgorithm)
+            {
+                int? alg = DecodeCoseAlgorithmHeader(encodedAlg);
+                Debug.Assert(alg.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+
+                if (expectedAlg != alg.Value)
+                {
+                    Type typeOfKey;
+                    if (keyType == KeyType.ECDsa)
+                    {
+                        typeOfKey = typeof(ECDsa);
+                    }
+                    else
+                    {
+                        typeOfKey = typeof(RSA);
+                    }
+
+                    throw new CryptographicException(SR.Format(SR.Sign1SignCoseAlgorithDoesNotMatchSpecifiedKeyAndHashAlgorithm, alg.Value, typeOfKey, hashAlgorithm.Name));
```

Comment:
This is building a System.Type when only a string is needed.

```suggestion
                    string typeOfKey = keyType switch
                    {
                        KeyType.ECDsa => "ECDSA",
                        KeyType.RSA => "RSA",
                        _ => throw new InvalidOperationException(),
                    };

                    throw new CryptographicException(SR.Format(SR.Sign1SignCoseAlgorithDoesNotMatchSpecifiedKeyAndHashAlgorithm, alg.Value, typeOfKey, hashAlgorithm.Name));
```

or even just use the enum ToString:

```suggestion
                    throw new CryptographicException(SR.Format(SR.Sign1SignCoseAlgorithDoesNotMatchSpecifiedKeyAndHashAlgorithm, alg.Value, keyType, hashAlgorithm.Name));
```

Reply:
I wanted the nice "System.Security.Cryptography.ECDsa" type to be printed but I think I can live with just "ECDsa".